File: 24th April - Nearest smaller tower/24thApril.java
class Solution{
	int [] nearestSmallestTower(int [] arr){
		//Write your code here
		int n = arr.length;
		Stack <Integer> stack = new Stack<>();
		
		int pre[] = new int[n];
		int suf[] = new int[n];
		
		// from left side
		for(int i = 0; i<n; i++){

		    while(!stack.isEmpty() && arr[stack.peek()]>= arr[i]) stack.pop();
		    
		    if(stack.isEmpty()) pre[i] = -1;
		    else pre[i] = stack.peek();
		    
		    stack.push(i);
		}
		
		//from right side
		for(int i = n-1; i>=0; i--){

		    while(!stack.isEmpty() && arr[stack.peek()]>= arr[i]) stack.pop();
		    
		    if(stack.isEmpty()) suf[i] = -1;
		    else suf[i] = stack.peek();
		    
		    stack.push(i);
		}
		
		int ans[] = new int[n];
		
		// answer
		for(int i = 0; i<n; i++){
		    if(pre[i]== -1 && suf[i]==-1) ans[i] = -1;
		    else if(pre[i] == -1 ) ans[i] = suf[i];
		    else if(suf[i] == -1) ans[i] = pre[i];
	        else if(i - pre[i] == suf[i] - i) {
	            if(arr[pre[i]]<=arr[suf[i]]) ans[i] = pre[i];
	            else ans[i] = suf[i];
	        }
		    else if(i - pre[i] > suf[i] - i) ans[i] = suf[i];
		    else ans[i] = pre[i];
		}
		return ans;
	}
	

}

File: April 2023/24th April - Nearest smaller tower/24thApril.py
class Solution:
    def nearestSmallestTower(self,arr):
        #code here
        n = len(arr)
        stack, smallest_to_left, smallest_to_right = [], [], []

        for i in range(n):
            while len(stack) > 0 and arr[stack[-1]] >= arr[i]:
                stack.pop()
                
            if len(stack) == 0:
                smallest_to_left.append(-1)
            else:
                smallest_to_left.append(stack[-1])
            stack.append(i)
        
        stack.clear()    
            
        for i in range(n-1, -1, -1):
            while len(stack) > 0 and arr[stack[-1]] >= arr[i]:
                stack.pop()
                
            if len(stack) == 0:
                smallest_to_right.append(-1)
            else:
                smallest_to_right.append(stack[-1])
            stack.append(i)
        smallest_to_right = smallest_to_right[::-1]
            
        res = []
        
        for i in range(n):
            if smallest_to_left[i] == smallest_to_right[i]:
                res.append(-1)
            elif smallest_to_left[i] == -1:
                res.append(smallest_to_right[i])
            elif smallest_to_right[i] == -1:
                res.append(smallest_to_left[i])
            
            elif i - smallest_to_left[i] == smallest_to_right[i] - i:
                if arr[smallest_to_left[i]] < arr[smallest_to_right[i]]:
                    res.append(smallest_to_left[i])
                elif arr[smallest_to_left[i]] == arr[smallest_to_right[i]]:
                    res.append(min(smallest_to_left[i], smallest_to_right[i]))
                else:
                    res.append(smallest_to_right[i])
                    
            else:
                if i - smallest_to_left[i] < smallest_to_right[i] - i:
                    res.append(smallest_to_left[i])
                else:
                    res.append(smallest_to_right[i])
                    
                    
        return res